doc/release/1.14.0-notes.rst

```diff
@@ -183,6 +183,14 @@ raising a ``TypeError``.
 -----------------------------------------------------------------
 When indexed with a float, the dtype object used to raise ``ValueError``.
 
+``nditer(... ['updateifcopy'], ...)`` must be used in a context manager
+---------------------------------------------------------------------
+Under certain conditions, calls to ``nditer()`` could create operands with
+writebackoncopy semantics. The resolution was triggerd in the ``nditer``
+``dealloc`` call, which let to strange ``i = None`` code after iteration to
+magically write the temporary data back to the original source. Now nditers
+can be used as context managers, and when writebackoncopy semantics are
+triggered they *must* be used as context managers.
```

Comment:
Unfortunately, AFAICT there are circumstances where `nditer` can use UPDATEIFCOPY even if `"updateifcopy"` isn't passed, like if you have a writable array that overlaps with another array.

Suggested text:

```rst
Under certain conditions, nditer must be used in a context manager
-----------------------------------------------
When using an nditer with the `"writeonly"` or `"readwrite"` flags, there are some circumstances where nditer doesn't actually give you a view onto the writable array. Instead, it gives you a copy, and if you make changes to the copy, nditer later writes those changes back into your actual array. Currently, this writeback depends on when the array objects are garbage collected, which makes this API error-prone on CPython and entirely broken on PyPy. Therefore, `nditer` can now be used as a context manager (`with np.nditer(...) as it: ...`), and in the future this will be required whenever using `nditer` with writable arrays.
```

Reply:
Note also that I'm suggesting here that we actually make the criterion for issuing deprecation warnings (now) and raising an error (later) be no context manager + writable arrays, since that's so much easier to explain to users than the actual edge cases. (And in particular if you have writable arrays then it's easy to get into a situation where in testing you never hit the UPDATEIFCOPY path, but then in production...)

Reply:
I tried to couch the wording in a way that did not limit it to the single flag. Requiring any writable nditer to be used inside a context manager pretty much means any nditer must be used that way. Is a change that deep acceptable/desired?

Reply:
The problem is that anyone who uses writable nditer *might* need this depending on details of the input arrays memory layout (in particular whether they overlap). I think forcing everyone to change at once with a nice warning message is better than forcing people to change piecemeal as they accidentally hit the edge case one by one.

Fortunately almost no-one uses nditer :-)

Reply:
I went through all the hits for `nditer` on searchcode.com, and only found ~15 projects, of which only ~5 use writeonly/readwrite. I'm sure there are others not in their corpus, but we're still only talking about probably ~dozens of pieces of code worldwide that need updating.

Reply:
fixed in 7eb8aed72